feat: add array builtin test scripts for E2E testing

Unisay opened the pull request:
## Context

- **What**: Implement complete array builtin test coverage for Protocol Version 11
- **Why**: CIP-0138 array builtins need E2E validation across all Plutus version combinations
- **Issue**: Resolves #14

## Approach

Implemented comprehensive test scripts for the three array builtin functions introduced in CIP-0138 (`indexArray`, `lengthOfArray`, `listToArray`) spanning the full Plutus version matrix.

**Version Matrix Coverage:**
- 3 array operations
- × 3 Plutus ledger versions (V1, V2, V3)
- × 2 Plutus Core language versions (1.0.0, 1.1.0)
- = **18 total test scripts**

**Naming Convention (V{Ledger}_{CoreCompact}):**

Established a new versioning naming convention that provides clear separation between version concepts:
- **Underscore** separates Plutus ledger version from Core version
- **No separator** within Core version digits (compacted: 110, 100)

Examples:
- `V3_110` = PlutusV3 + Plutus Core 1.1.0
- `V1_100` = PlutusV1 + Plutus Core 1.0.0

This convention:
- Removes ambiguity about which separator indicates what
- Provides visual clarity: `V3_110` vs verbose `V3_1_1_0`
- Makes grepping easier: `V3_*` finds all V3, `*_110` finds all 1.1.0
- Sets precedent for other batch6 builtins (dropList, etc.)

**Architecture:**

Follows established project pattern:
- **Common.hs**: INLINEABLE validator logic shared across versions
- **Versioned modules** (V1_100, V1_110, V2_100, V2_110, V3_100, V3_110): Compile scripts with specific version targets
- **app/Main.hs**: Serializes all scripts to `.plutus` envelopes
- **CompiledCode exports**: Pure compilation, no IO in library

Each validator tests one array builtin operation with hardcoded test data:
- `indexArray`: Access element at index 2 from [1,2,3,4,5], expects 3
- `lengthOfArray`: Get length of [10,20,30], expects 3
- `listToArray`: Convert [100,200,300,400], verify length and boundary elements

## Changes

### Core Implementation

**PlutusScripts/Array/Common.hs**
- `mkIndexArrayPolicy`: Tests `indexArray` builtin with valid index access
- `mkLengthOfArrayPolicy`: Tests `lengthOfArray` builtin
- `mkListToArrayPolicy`: Tests `listToArray` conversion with verification
- Uses `P.toOpaque` for list construction (PlutusTx plugin requirement)
- Returns `BI.unitval` on success, `P.traceError` on failure

**Versioned Compilation Modules (6 files)**
- V1_100.hs: PlutusV1 + Core 1.0.0
- V1_110.hs: PlutusV1 + Core 1.1.0
- V2_100.hs: PlutusV2 + Core 1.0.0
- V2_110.hs: PlutusV2 + Core 1.1.0
- V3_100.hs: PlutusV3 + Core 1.0.0
- V3_110.hs: PlutusV3 + Core 1.1.0

Each exports:
- `succeedingIndexArrayPolicyCompiled_V{X}_{YZ0}` (CompiledCode)
- `succeedingLengthOfArrayPolicyCompiled_V{X}_{YZ0}` (CompiledCode)
- `succeedingListToArrayPolicyCompiled_V{X}_{YZ0}` (CompiledCode)

## Generated Scripts (18 total)

### PlutusV1 (6 scripts)
- `succeedingIndexArrayPolicyScript_V1_100.plutus` (339 bytes)
- `succeedingLengthOfArrayPolicyScript_V1_100.plutus` (318 bytes)
- `succeedingListToArrayPolicyScript_V1_100.plutus` (410 bytes)
- `succeedingIndexArrayPolicyScript_V1_110.plutus` (305 bytes)
- `succeedingLengthOfArrayPolicyScript_V1_110.plutus` (292 bytes)
- `succeedingListToArrayPolicyScript_V1_110.plutus` (366 bytes)

### PlutusV2 (6 scripts)
- `succeedingIndexArrayPolicyScript_V2_100.plutus` (339 bytes)
- `succeedingLengthOfArrayPolicyScript_V2_100.plutus` (318 bytes)
- `succeedingListToArrayPolicyScript_V2_100.plutus` (410 bytes)
- `succeedingIndexArrayPolicyScript_V2_110.plutus` (305 bytes)
- `succeedingLengthOfArrayPolicyScript_V2_110.plutus` (292 bytes)
- `succeedingListToArrayPolicyScript_V2_110.plutus` (366 bytes)

### PlutusV3 (6 scripts)
- `succeedingIndexArrayPolicyScript_V3_100.plutus` (339 bytes)
- `succeedingLengthOfArrayPolicyScript_V3_100.plutus` (318 bytes)
- `succeedingListToArrayPolicyScript_V3_100.plutus` (410 bytes)
- `succeedingIndexArrayPolicyScript_V3_110.plutus` (305 bytes)
- `succeedingLengthOfArrayPolicyScript_V3_110.plutus` (292 bytes)
- `succeedingListToArrayPolicyScript_V3_110.plutus` (366 bytes)

## References

- Resolves #14
- CIP-0138: https://cips.cardano.org/cip/CIP-0138
- Plutus conformance tests: https://github.com/IntersectMBO/plutus/tree/master/plutus-conformance/test-cases/uplc/evaluation/builtin/semantics
- Related: PR #15 (applies same naming to dropList)

Unisay committed:
feat: add array builtin test scripts for E2E testing

Add three PlutusV3 scripts for testing array builtins (CIP-0138):
- succeedingIndexArrayPolicyScriptV3: Tests indexArray builtin
- succeedingLengthOfArrayPolicyScriptV3: Tests lengthOfArray builtin
- succeedingListToArrayPolicyScriptV3: Tests listToArray builtin

Implementation follows existing bitwise success script pattern with
validators using Plutus Core 1.1.0 target version.

Changes:
- Add PlutusScripts/Array/Common.hs with INLINEABLE validator logic
- Add PlutusScripts/Array/V_1_1.hs with PlutusV3 compilations
- Update app/Main.hs to generate array script envelopes
- Update plutus-scripts.cabal with new Array modules

Unisay committed:
feat: expand array builtins to full Plutus version matrix (issue #14)

Implement complete array builtin test coverage for Protocol Version 11:
- 3 functions × 3 Plutus versions × 2 language versions = 18 scripts total

Created 6 versioned modules:
- V1_1_0_0.hs (PlutusV1/1.0.0)
- V1_1_1_0.hs (PlutusV1/1.1.0)
- V2_1_0_0.hs (PlutusV2/1.0.0)
- V2_1_1_0.hs (PlutusV2/1.1.0)
- V3_1_0_0.hs (PlutusV3/1.0.0)
- V3_1_1_0.hs (PlutusV3/1.1.0) - renamed from V_1_1.hs

Generated Scripts (18 total):
V1/1.0.0: indexArray, lengthOfArray, listToArray
V1/1.1.0: indexArray, lengthOfArray, listToArray
V2/1.0.0: indexArray, lengthOfArray, listToArray
V2/1.1.0: indexArray, lengthOfArray, listToArray
V3/1.0.0: indexArray, lengthOfArray, listToArray
V3/1.1.0: indexArray, lengthOfArray, listToArray

Changes:
- Add 5 new versioned modules + rename existing V_1_1.hs
- Update plutus-scripts.cabal with all 6 versioned modules
- Add writeEnvelopeV1 helper in app/Main.hs
- Generate all 18 script envelopes for E2E testing

Closes #14

Unisay committed:
refactor: adopt compacted Core version naming (V3_110)

Change naming convention to compact Plutus Core version digits
for clearer separation between version concepts and version digits.

Convention: V{LedgerVersion}_{CoreCompact}
- Underscore separates distinct concepts (Plutus vs Core version)
- No separator within Core version digits (compacted: 110, 100)

Examples:
- V3_110 = PlutusV3 + Plutus Core 1.1.0
- V1_100 = PlutusV1 + Plutus Core 1.0.0

Changes:
- Rename 6 modules: V{X}_{Y}_{Z}_{W}.hs → V{X}_{YZW}.hs
- Update all identifiers: ...V3_1_1_0 → ..._V3_110
- Update script outputs: ...V3_1_1_0.plutus → ..._V3_110.plutus
- Update qualified imports: Array_V3_110 (cleaner aliases)
- Document convention in CLAUDE.md

Generated scripts (18 total):
- succeedingIndexArrayPolicyScript_V{1,2,3}_{100,110}.plutus
- succeedingLengthOfArrayPolicyScript_V{1,2,3}_{100,110}.plutus
- succeedingListToArrayPolicyScript_V{1,2,3}_{100,110}.plutus

Benefits:
- Less verbose: V3_110 vs V3_1_1_0 (6 vs 8 chars)
- Clearer hierarchy: version concepts separated, digits compact
- Better readability and grep patterns

Related: issue #14

kwxm commented on plutus-scripts/PlutusScripts/Batch6/Array/Common.hs:
```diff
@@ -0,0 +1,45 @@
+)
+where
+
+import PlutusTx.Builtins qualified as BI
```
I guess there's no reason why all of the builtins have their own scripts.  In this case we could have a single script that carries out some task that uses all of the builtins, and that would reduce the number of scripts we have tomanage, although doing that might make it tricky to pin down what's causing the problem if the test fails.  That's just an observation though: there's no need to change these.

kwxm commented on plutus-scripts/app/Main.hs:
```diff
@@ -49,6 +46,54 @@ main = withUtf8 do
+    Array_V3_100.succeedingListToArrayPolicyCompiled_V3_100
+
+  -- Array builtin scripts (V3/1.1.0)
+  writeEnvelopeV3 "succeedingIndexArrayPolicyScript_V3_110"
```
Would it simplify things to have a list of pairs exported from the source files for the different versions and map `writeEnvelope` over that?  We're going to end up with tests for fourteen new builtins and things may get a little out of hand.  Alternatively, have a `WriteEnvelopes` file (or something similar) for each builtin or set of builtins and call some function from that here to write all of the scripts.

kwxm reviewed (changes requested):
As I said in the other PR, I think we should probably make a `Batch6` directory in `PlutusScripts` and put all the array/dropList/expModInteger/BLS-MSM/built-in-value tests in there.

kwxm commented:
Oh yes, if you feel like it you could maybe add a failing test for out-of-bounds array access.

Unisay committed:
refactor: move array builtins to Batch6/Array/ subdirectory

Reorganize array builtins under Batch6 directory for protocol version
tracking, following Kenneth's feedback to group all batch6 builtins.

Directory structure:
  Batch6/
  └── Array/
      ├── Common.hs
      ├── V1_100.hs
      ├── V1_110.hs
      ├── V2_100.hs
      ├── V2_110.hs
      ├── V3_100.hs
      └── V3_110.hs

Changes:
- Move PlutusScripts/Array/ → PlutusScripts/Batch6/Array/
- Update module paths: Batch6.Array.V3_110
- Update imports in all versioned modules
- Update cabal exposed-modules
- Update Main.hs imports

This matches the Batch6/DropList/ structure (PR #15) and enables
clean organization of all batch6 builtins in one location.

Generated scripts (18 total):
- succeedingIndexArrayPolicyScript_V{1,2,3}_{100,110}.plutus
- succeedingLengthOfArrayPolicyScript_V{1,2,3}_{100,110}.plutus
- succeedingListToArrayPolicyScript_V{1,2,3}_{100,110}.plutus

Related: PR #15 (dropList), Kenneth's feedback, issue #14

Unisay committed:
docs: update CLAUDE.md examples for Batch6 subdirectory structure

Update examples to reflect reorganization:
- PlutusScripts.Batch6.Array.V3_110
- PlutusScripts.Batch6.DropList.V3_110

This matches the actual directory structure after reorganization.

Unisay committed:
refactor: simplify identifiers by removing version repetition

Remove version suffix from identifiers to avoid repetition:
- Old: Array_V2_100.succeedingIndexArrayPolicyCompiled_V2_100
- New: Array_V2_100.succeedingIndexArrayPolicy

Changes:
- Remove version suffix from all exported identifiers
- Remove SerialisedScript exports (only export CompiledCode)
- Remove unused imports
- Version info already clear from qualified import

Example usage:
  Array_V3_110.succeedingIndexArrayPolicy
  Array_V1_100.succeedingLengthOfArrayPolicy

Benefits: cleaner code, no excessive repetition, version still clear.

Unisay committed:
refactor: introduce VersionedScript pattern for batch6 builtins

Replace repetitive writeEnvelope calls with structured VersionedScript
list pattern, reducing 42 lines to 1 line for array builtins.

New pattern:
- Created Helpers/Envelopes with VersionedScript type
- Created PlutusScripts/Batch6/Array overarching module
- Exports allArrayScripts list with version metadata
- Main.hs: mapM_ writeVersionedScript Array.allArrayScripts

Benefits:
- Main.hs scales: 1 line per builtin (not 3N lines for N operations)
- No version repetition: Array_V2_100.succeedingIndexArrayPolicy
- Type-safe version tracking with proper PlutusCore type
- Easy to extend: add operation = add one entry to list

Example:
  Old (42 lines):
    writeEnvelopeV1 "succeedingIndexArrayPolicyScript_V1_100"
      Array_V1_100.succeedingIndexArrayPolicy
    ... (repeated 18 times)

  New (1 line):
    mapM_ writeVersionedScript Array.allArrayScripts

With 14 batch6 builtins, Main.hs stays clean and maintainable.

Related: Kenneth's feedback on PR #12

kwxm reviewed (approved):
This looks great: thanks!
